Refactor reward logging to use running totals

nathandsouza10 opened the pull request:
- Removed list-based reward accumulation
- Implemented running total/count for individual and group rewards
- Streamlined TensorBoard logging to use calculated averages

nathandsouza10 committed:
Refactor reward logging to use running totals

- Removed list-based reward accumulation
- Implemented running total/count for individual and group rewards
- Streamlined TensorBoard logging to use calculated averages

jaredvann commented:
@nathandsouza10 Hi, thank you for taking the time to create this PR. We are not completely clear on what the overall benefits of these changes are. Is the goal optimisation of memory usage from not needing to store lists of reward values?

nathandsouza10 commented:
> @nathandsouza10 Hi, thank you for taking the time to create this PR. We are not completely clear on what the overall benefits of these changes are. Is the goal optimisation of memory usage from not needing to store lists of reward values?

Hi, thank you for considering my PR. Yes, the primary goal of these changes, albeit small, is to optimize memory usage. By moving away from storing extensive lists of reward values and implementing a running total and count for each agent, we significantly reduce the memory overhead. This is particularly beneficial in scenarios with many agents or long-running environments.

jaredvann commented on phantom/trainer.py:
```diff
@@ -226,7 +226,10 @@ def log_vec_metrics(self, envs: Sequence[PhantomEnv]) -> None:
         """Logs the rewards from a provided env."""
         for agent_id, reward in rewards.items():
-            self.logged_rewards[agent_id].append(reward)
+            if agent_id not in self.logged_rewards:
+                self.logged_rewards[agent_id] = {"total": 0, "count": 0}
```
This should not be necessary because of the defaultdict default value in line 255. Please can you test without this.

nathandsouza10 replied:
Thank you for highlighting the redundancy in the `defaultdict` usage. I believe you are correct; the explicit initialization of `self.logged_rewards[agent_id]` is indeed unnecessary due to the `defaultdict`'s behavior.

jaredvann commented:
Thanks! Please see one small comment below.

To be able to contribute to this repo and merge this PR you will need to first sign the JPMorgan Chase Contributor License Agreement (CLA). This can be found at https://github.com/jpmorganchase/.github/blob/main/jpmc-cla-20230406.md. Instructions can be found at the bottom of the file.

Please let me know if you have any questions.

nathandsouza10 committed:
Merge branch 'develop' into refactor/optimize-reward-logging

nathandsouza10 commented:
> any questions.

Thank you for guiding me through the contribution process. I wanted to inform you that I have already completed the submission of the JPMorgan Chase Contributor License Agreement (CLA). This was done on Friday, December 5th, and I sent it to opensource_cla@jpmchase.com. Please do let me know if there is anything else I need to do or if there are any further steps to finalize my contribution. I’m looking forward to actively participating in the project!

leoardon commented:
@nathandsouza10 Thanks for raising the request to contribute, it might need some time to be reviewed and accepted. We will follow up on this. 
Could you please make the change mentioned and make sure it works as expected?

nathandsouza10 committed:
Remove redundant initialization of logged_rewards

Following review feedback, the explicit initialisation of logged_rewards[agent_id] has been removed.

nathandsouza10 committed:
Format code using black for workflow compliance

leoardon commented:
@nathandsouza10 Good news, I see your CLA request has gone through and you're now an official contributor! :tada: ✨ 

Could you also add unit tests to make sure your changes are working please? I see for example that the `logged_rewards` are still initialized as a `list` in the `__init__` method l.89.
